pkg/bgpv1/agent/controller.go

```diff
@@ -221,33 +250,45 @@ func NewController(ctx context.Context, clientset client.Clientset, rtMgr BGPRou
 	}
 
 	// setup policy lister and informer.
-	policyLister := factory.Cilium().V2alpha1().CiliumBGPPeeringPolicies().Lister()
-	policyInformer := factory.Cilium().V2alpha1().CiliumBGPPeeringPolicies().Informer()
-	policyInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+	c.PolicyLister = factory.Cilium().V2alpha1().CiliumBGPPeeringPolicies().Lister()
+	c.policyInformer = factory.Cilium().V2alpha1().CiliumBGPPeeringPolicies().Informer()
+	c.policyInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc:    sig.Event,
 		UpdateFunc: func(_ interface{}, _ interface{}) { sig.Event(struct{}{}) },
 		DeleteFunc: sig.Event,
 	})
-	go policyInformer.Run(stop)
-
-	c := Controller{
-		NodeSpec:     nodeSpecer,
-		PolicyLister: policyLister,
-		Sig:          sig,
-		BGPMgr:       rtMgr,
-	}
 
 	// apply any options.
-	for _, opt := range opts {
+	for _, opt := range params.Opts {
 		opt(&c)
 	}
 
-	// stop channel is passed here, if ctx is canceled the passed stop chan
-	// is closed as well.
-	go c.Run(ctx, stop)
+	params.LC.Append(&c)
+
 	return &c, nil
 }
 
+// Start is called by hive after all of our dependencies have been started.
+func (c *Controller) Start(_ hive.HookContext) error {
+	go c.policyInformer.Run(c.ctx.Done())
+	go c.NodeSpec.Run(c.ctx)
+	go c.Run()
```

Comment:
My gut wants Run() to take a ctx here.

Reply:
Yea, this again comes down to `context.Context` usage patterns. I know you prefer passing via parameter, so I will refactor to that pattern.

Reply:
Gracias.